### PR TITLE
Fix MagicNumber rule to support suppression of negative constants in ranges (#8338)

### LINKED ISSUE #8338: MagicNumber suppression for ranges does not work for negative numbers
In #1368 a configuration was added to allow suppressing MagicNumber for ranges

## Expected Behavior
That works fine for positive values but does not work for negative values. It should not report Magic Number for ranges that involve negative numbers

## Observed Behavior
```
val doesNotReportMagicNumber = 5..10
val doesReportMagicNumber = -15..-10
```

## Context
Requires me to suppress MagicNumber

### PR DESCRIPTION
**Pull request by sammetav**
Fix MagicNumber rule to support suppression of negative constants in ranges (#8338)

* Fix MagicNumber rule to correctly suppress negative values in ranges

- Enhance `isPartOfRange()` to detect negative constants wrapped in `KtPrefixExpression`
- Add support for cases like `-15..-10`, `-20 downTo -25`, etc.

Fixes #8338

* test: add test case for negative range expression

### CODE REVIEW

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt [hunk 1]
```diff
@@ -197,12 +197,22 @@ class MagicNumber(config: Config) : Rule(
...
+        val isDirectRange = theParent is KtBinaryExpression &&
+            (
+                theParent.operationToken == KtTokens.RANGE ||
+                    theParent.operationReference.getReferencedName() in rangeOperators
+                )
+
+        val isNegativePartOfRange = theParent is KtPrefixExpression &&
+            (theParent.firstChild as? KtOperationReferenceExpression)?.operationSignTokenType == KtTokens.MINUS &&
+            theParent.parent is KtBinaryExpression &&
+            (theParent.parent as KtBinaryExpression).let { grandParent ->
```
Comment (harrytmthy) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt:210]: NIT: Would it be better to avoid repeated casts and separate the two cases?

```kotlin
private fun KtConstantExpression.isPartOfRange(): Boolean {
    val theParent = parent
    val rangeOperators = setOf("downTo", "until", "step")
    
    // Direct range: 1..10, 1 downTo 0
    if (theParent is KtBinaryExpression) {
        return theParent.operationToken == KtTokens.RANGE ||
               theParent.operationReference.getReferencedName() in rangeOperators
    }
    
    // Negative in range: (-5)..10
    if (theParent is KtPrefixExpression &&
        theParent.operationReference?.operationSignTokenType == KtTokens.MINUS
    ) {
        val grandParent = theParent.parent as? KtBinaryExpression ?: return false
        val isRangeOperation = grandParent.operationToken == KtTokens.RANGE ||
                               grandParent.operationReference.getReferencedName() in rangeOperators
        
        return isRangeOperation
    }
    
    return false
}
```

This retains the existing style and avoids duplicate casting on `theParent.parent`.

Comment (harrytmthy) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt]: NIT: Would it be better to avoid repeated casts and separate the two cases?

```kotlin
private fun KtConstantExpression.isPartOfRange(): Boolean {
    val theParent = parent
    val rangeOperators = setOf("downTo", "until", "step")
    
    // Direct range: 1..10, 1 downTo 0
    if (theParent is KtBinaryExpression) {
        return theParent.operationToken == KtTokens.RANGE ||
               theParent.operationReference.getReferencedName() in rangeOperators
    }
    
    // Negative in range: (-5)..10
    if (theParent is KtPrefixExpression &&
        theParent.operationReference?.operationSignTokenType == KtTokens.MINUS
    ) {
        val grandParent = theParent.parent as? KtBinaryExpression ?: return false
        return grandParent.operationToken == KtTokens.RANGE ||
               grandParent.operationReference.getReferencedName() in rangeOperators
    }
    
    return false
}
```

This retains the existing style and avoids duplicate casting on `theParent.parent`.

Comment (sammetav) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt]: IIRC I ran into operationSignTokenType being unavailable directly on operationReference, so a cast is still needed, but of course we can change the structure to retain the existing style.

```
 @Suppress("ReturnCount")
    private fun KtConstantExpression.isPartOfRange(): Boolean {
        val theParent = parent
        val rangeOperators = setOf("downTo", "until", "step")

        // Case 1: Direct range expression
        if (theParent is KtBinaryExpression) {
            return theParent.operationToken == KtTokens.RANGE ||
                theParent.operationReference.getReferencedName() in rangeOperators
        }

        // Case 2: Negative number part of a range
        if (theParent is KtPrefixExpression) {
            val opRef = theParent.firstChild as? KtOperationReferenceExpression
            if (opRef?.operationSignTokenType != KtTokens.MINUS) return false

            val grandParent = theParent.parent as? KtBinaryExpression ?: return false
            return grandParent.operationToken == KtTokens.RANGE ||
                grandParent.operationReference.getReferencedName() in rangeOperators
        }

        return false
    }
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumberSpec.kt [hunk 2]
```diff
@@ -920,6 +920,12 @@ class MagicNumberSpec {
         }
     }
 
+    @Test
+    fun `does not report negative numbers in ranges when ranges are ignored`() {
+        val code = "val range = -15..-10"
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumberSpec.kt]: You could add cases to the `cases` function on top. I think that would make the tests easier to maintain and get better coverage.

**Review by BraisGabin — COMMENTED** [review]
Thanks for the fix. I left a comment about the tests. Also you say that you spot `..` and `until`. Add a case for until too please.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt (hunk 1 above)
Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt]: I like the mulple ifs. It should be faster. Also we could move `rangeOperators` to a companion object so it is not created over and over again.

Comment (sammetav) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/MagicNumber.kt]: 👌
